### PR TITLE
Issue#19 corrup star file handling

### PR DESCRIPTION
**Pull request by serpens24**
### Description of the Change

Added an additional level of redundacy to star file loads, such that backup files (.bak) are created on save, prior to saving of new file verison. These backup files capture the star file before updates, (because these files were able to be loaded). Additionally, no longer delete backup file when loading an autosave file .

Upon load, andattmept is made to open the selected star file, if this fails, a check is made to see if an associated backup is found and if sao, attempts are made to load this. 

### Alternate Designs

I considered options to try and capture all ways a file could be corrupted, but there are too many to consider, hence is is considered easier to be more robust in loading.

### Why Should This Be In Core?

Bug fix.

### Benefits

Add an extra level of redundancy to file loading.

### Possible Drawbacks

There is one case still not covered, if an autosave file exists and actual file is corrupted.

### Verification Process

Us the following table: 
image

Create a set of valid and invalid files. Itertate through the scenarios identified in the above table and confirm that the file loaded (or error) is in line with table.

### Applicable Issues

#19

### CODE REVIEW

File: CoreGraphFile/src/au/gov/asd/tac/constellation/graph/file/io/GraphJsonReader.java [hunk 1]
```diff
@@ -100,15 +100,20 @@ public GraphJsonReader() {
...
     }
 
     public Graph readGraphZip(final String name, InputStream bin, final IoProgress progress) throws IOException, GraphParseException {
         try (bin) {
             progress.start(100);
-            progress.progress("Reading file: " + name);
-            
             byteReader = new GraphByteReader(bin);
+        } 
+        catch (IOException ex) {
```
Comment (antares1470) [inline CoreGraphFile/src/au/gov/asd/tac/constellation/graph/file/io/GraphJsonReader.java:112]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java [hunk 2]
```diff
@@ -176,12 +178,45 @@ protected Void doInBackground() throws Exception {
...
                     gex = ex;
                 }
 
+                try {
+                    if (gex != null) {
+                        // An exception was thrown trying to read specified star file. The most likely reason for this is
+                        // a corrupt star file. Check to see if there was a 'backup' star file generated before the star file
+                        // was writtien - if so, attmept to load this.
+                        String backupFilename = graphFile.toString().concat(BACKUP_EXTENSION);
+                        File backupFile = new File(backupFilename);
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:194]: Both these can be made final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java [hunk 3]
```diff
@@ -176,12 +178,45 @@ protected Void doInBackground() throws Exception {
             if (graph == null) {
                 try {
                     final long t0 = System.currentTimeMillis();
+                    LOGGER.log(Level.INFO, String.format("MMDEBUG Attempting to open %s", graphFile.toString()));
                     graph = new GraphJsonReader().readGraphZip(graphFile, new HandleIoProgress(String.format("Reading %s...", graphFile.getName())));
                     time = System.currentTimeMillis() - t0;
-                } catch (GraphParseException | IOException | RuntimeException ex) {
+                } catch (GraphParseException | IOException | RuntimeException ex) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:184]: Noting you didn't add this but if could make this final as well that would be great

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java [hunk 4]
```diff
@@ -176,12 +178,45 @@ protected Void doInBackground() throws Exception {
...
+                            
+                            // Backup file successfully loaded, copy it over top of corrupt actual file - theres no reason to keep the corrupted file.
+                            // Don't do a move, rather perform the move in two stages, a copy, then a delete to ensure there
+                            // is always going to be a valid file somewhere as only the copy or the delete can fail in a given run.
+                            FileUtils.copyFile(new File(backupFile.toString()), new File(graphFile.toString()));
+                            FileUtils.deleteQuietly(backupFile);
+                        }
+                    }
+                }
+                catch (GraphParseException | IOException | RuntimeException ex) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:215]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java [hunk 5]
```diff
@@ -176,12 +178,45 @@ protected Void doInBackground() throws Exception {
             if (graph == null) {
                 try {
                     final long t0 = System.currentTimeMillis();
+                    LOGGER.log(Level.INFO, String.format("MMDEBUG Attempting to open %s", graphFile.toString()));
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:181]: I suspect some of these debugging log entries can be removed

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java [hunk 6]
```diff
@@ -1093,6 +1095,29 @@ protected Void doInBackground() throws Exception {
...
+                //     have something to fall back on
+                // This allows load logic to be of the form:
+                //   - try to load star file
+                //     - if star file loads successfully use it
+                //     - if the star file fails to load, check to see if a 'backup' exists
+                //        - if backup was found attempt to load it
+                //        - if backup was not found throw load error
+                FileObject fileobj = freshGdo.getPrimaryFile();
+                File srcFile = new File(fileobj.getPath());
+                String srcfilePath = srcFile.getParent().concat("\\").concat(this.name).concat(".").concat(fileobj.getExt());
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java:1114]: These 3 can be final

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Comment by arcturus2** [comment]
@serpens24 I started by doing a normal graph create and then closed Constellation by stopping the process I am left with the `AutoSave` files left around.

So 2 are small graphs and 1 was a 10k node and transaction graph. When I loaded Constellation it detected the 2 small graphs but not the 3rd larger one.

image

Closing and opening Constellation again normally did not detect anything in `Autosave` and the files are still sticking around. Is that expected?

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java (hunk 6 above)
Comment (arcturus2) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java:1114]: Should the `\\` be replaced with `File.pathSeparator` so we support Linux and Mac?

Comment (aldebaran30701) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java:1114]: `SeparatorConstants` might have a `.PERIOD` also

**Review by aldebaran30701 — COMMENTED** [review]
Tested functionality, and it behaves correctly as described. 
Happy to approve this once the fixes from antares and arcturus' review are finished.
I tried to replicate what @arcturus2 has explained about the Autosave files. Mine seems to have 3 autosave files (6 total) and it offers 3 recent autosaves as intended.

File: CoreGraphFile/src/au/gov/asd/tac/constellation/graph/file/io/GraphJsonReader.java (hunk 1 above)
Comment (serpens24) [reply CoreGraphFile/src/au/gov/asd/tac/constellation/graph/file/io/GraphJsonReader.java:112]: addressed, also checked for others in file.

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java (hunk 2 above)
Comment (serpens24) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:194]: addressed - also looked for others in file.

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java (hunk 3 above)
Comment (serpens24) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:184]: as above

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java (hunk 4 above)
Comment (serpens24) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphOpener.java:215]: as above

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java (hunk 6 above)
Comment (serpens24) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java:1114]: done

Comment (serpens24) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/gui/VisualGraphTopComponent.java:1114]: done - for path seperatore .. .. left the . as its universal.

**Review by arcturus2 — APPROVED** [review]
Thank you for all the effort you have put into this change @serpens24. I have not been able to get to every scenario but the graphs seem to load OK from the autosave when Constellation was prematurely terminated.

One thing I'd note is that going back to issue #19, it could be a termination of Constellation whilst it is saving a file via File -> Save As. The star file was 0 bytes in size and I'm wondering if you looked into the JVM shutdown hook `Runtime.getRuntime().addShutdownHook()` already and what was the outcome from that? I think we previously discussed this but can't remember the exact outcome.

**Review by antares1470 — APPROVED** [review]

**Pull request merged**